src/servers/lsp.rs

```diff
@@ -390,6 +326,91 @@ impl Server for LspServer {
         self.publish_verify_statuses()?;
         Ok(())
     }
+
+    fn handle_not_checked(&mut self, span: Span) -> Result<(), ServerError> {
+        self.statuses.insert(span, VerifyResult::Unknown);
+        self.publish_verify_statuses()?;
+        Ok(())
+    }
+}
+
+pub async fn run_lsp_server(
+    server: Arc<Mutex<LspServer>>,
+    mut verify: impl FnMut(&[FileId]) -> Pin<Box<dyn Future<Output = Result<(), VerifyError>> + '_>>,
+) -> Result<(), VerifyError> {
+    let (sender, receiver) = {
+        let server_guard = server.lock().unwrap();
+        let sender = server_guard.connection.sender.clone();
+        let receiver = server_guard.connection.receiver.clone();
+        (sender, receiver)
+    };
+    for msg in &receiver {
+        match msg {
+            Message::Request(req) => match req.method.as_str() {
+                "custom/verify" => {
```

Comment:
Maybe also move this block to a separate function for clarity.

Reply:
And explain why we don't want/**must not** keep a lock on the server for the whole duration of the block.